src/docfx/lib/schema/JsonSchemaTransformer.cs

```diff
@@ -13,136 +14,146 @@ internal class JsonSchemaTransformer
+            var transformedToken = TransformToken(file, context, _schema, token, errors);
             return (errors, transformedToken);
         }
 
-        public (List<Error> errors, Dictionary<string, List<(bool isRoot, SourceInfo source, Dictionary<string, Lazy<JToken>> propertiesByUid)>> xrefPropertiesGroupByUid) TransformXref(Document file, Context context, JToken token)
+        public (List<Error> errors, Dictionary<string, List<(bool isRoot, SourceInfo source, Dictionary<string, Lazy<JToken>> propertiesByUid)>> xrefPropertiesGroupByUid) TraverseXref(Document file, Context context, JToken token)
         {
             var errors = new List<Error>();
             var xrefPropertiesGroupByUid = new Dictionary<string, List<(bool, SourceInfo, Dictionary<string, Lazy<JToken>>)>>();
 
-            Traverse(_schema, token, (schema, node) =>
+            TraverseXref(_schema, token);
+
+            return (errors, xrefPropertiesGroupByUid);
+
+            void TraverseXref(JsonSchema schema, JToken node)
             {
-                if (node is JObject obj)
+                schema = _definitions.GetDefinition(schema);
+                switch (node)
                 {
-                    var uid = obj.TryGetValue("uid", out var uidValue) && uidValue is JValue uidJValue && uidJValue.Value is string uidStr ? uidStr : null;
+                    case JObject obj:
+                        var uid = obj.TryGetValue("uid", out var uidValue) && uidValue is JValue uidJValue && uidJValue.Value is string uidStr ? uidStr : null;
 
-                    if (uid == null)
-                    {
-                        return (default, node);
-                    }
+                        if (uid is null)
+                        {
+                            TraverseObjectXref(obj);
+                            break;
+                        }
 
-                    if (!xrefPropertiesGroupByUid.TryGetValue(uid, out _))
-                    {
-                        xrefPropertiesGroupByUid[uid] = new List<(bool, SourceInfo, Dictionary<string, Lazy<JToken>>)> { (obj.Parent is null, JsonUtility.GetSourceInfo(obj), BuildXrefPropertiesForUid(schema, obj)) };
-                    }
-                    else
-                    {
-                        xrefPropertiesGroupByUid[uid].Add((obj.Parent is null, JsonUtility.GetSourceInfo(obj), BuildXrefPropertiesForUid(schema, obj)));
-                    }
+                        if (!xrefPropertiesGroupByUid.TryGetValue(uid, out var xrefPropertyList))
+                        {
+                            xrefPropertiesGroupByUid[uid] = xrefPropertyList = new List<(bool, SourceInfo, Dictionary<string, Lazy<JToken>>)>();
+                        }
 
-                    return (schema.XrefProperties, node);
+                        var xrefProperties = new Dictionary<string, Lazy<JToken>>();
+                        TraverseObjectXref(obj, (propertySchema, key, value) =>
```

Comment:
prefer calls over callbacks

Reply:
TraverseObjectXref need to be used in two places, just avoid duplicated code